pkg/store/bucket.go

```diff
@@ -804,8 +804,37 @@ func chunksSize(chks []storepb.AggrChunk) (size int) {
 }
 
 // LabelNames implements the storepb.StoreServer interface.
-func (s *BucketStore) LabelNames(context.Context, *storepb.LabelNamesRequest) (*storepb.LabelNamesResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+func (s *BucketStore) LabelNames(ctx context.Context, _ *storepb.LabelNamesRequest) (*storepb.LabelNamesResponse, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	s.mtx.RLock()
+
+	var mtx sync.Mutex
+	var sets [][]string
+
+	for _, b := range s.blocks {
+		indexr := b.indexReader(gctx)
+		g.Go(func() error {
+			defer runutil.CloseWithLogOnErr(s.logger, indexr, "label names")
+
+			res := indexr.LabelNames()
+
+			mtx.Lock()
+			sets = append(sets, res)
+			mtx.Unlock()
+
+			return nil
+		})
+	}
+
+	s.mtx.RUnlock()
+
+	if err := g.Wait(); err != nil {
+		return nil, status.Error(codes.Aborted, err.Error())
```

Comment:
Check here for partial response here and return it if it's enabled

Reply:
I have checked partial response at the `ProxyStore.LabelNames()`, should I do the same things here?

Reply:
I might have misunderstood things, the whole point is to make partialResponse work so once you finish the tests this might be not needed :)

Reply:
hm. I don't get you comment @povilasv either as literally there is no way this will return error :
```
defer runutil.CloseWithLogOnErr(s.logger, indexr, "label names")

 			res := indexr.LabelNames()

 			mtx.Lock()
			sets = append(sets, res)
			mtx.Unlock()
return nil
```

Reply:
so partial response does not exists here :+1:

Reply:
So the timeout case' context get's cancelled, not all blocks are `s.blocks ` are read, the `g.Wait()` returns an error, in this case we return Aborted, but might as well return partial data, no?

@bwplotka wdyt?

Reply:
@povilasv maybe `g.Wait()` will not return an error when the timeout context is cancelled. So in this case, there is no way to return errors in `g.Go()` block.

Reply:
So `g.Go()` needs to check `ctx.Err()`, before executing stuff.

```
	g.Go(func() error {
			defer runutil.CloseWithLogOnErr(s.logger, indexr, "label names")
                        if ctx.Err() != nil {
                        //handle partial data
                        }
 			res := indexr.LabelNames()

 			mtx.Lock()
			sets = append(sets, res)
			mtx.Unlock()

 			return nil
		})
```

Imagine the case where user canceld the request, and we still have tons of blocks to go thru.

Reply:
@jojohappy you forgot to address this ^

Reply:
Sorry, I had updated at this moment, I tried to check partial response enabled when `ctx` was timeout or cancelled.

Reply:
Quite confused by this thread, and sorry for delayed answer. I think partial response has really small reason of being handled here in store gw. It might just create confusion and being volation of YAGNI as no one requested every to ignore if block errors out. It is totally different to Querier partial response, where query might be still true if one StoreAPI is down